openff/toolkit/topology/molecule.py

```diff
@@ -4673,20 +4676,30 @@ def from_qcschema(
                 # try and find the initial molecule conformations and attach them
                 # collect the input molecules
                 try:
-                    input_mols = client.query_molecules(
-                        id=qca_record["initial_molecules"]
-                    )
+                    initial_molecules: list[dict] = qca_record["initial_molecules"]
+
+                    hashes: list[str] = [
+                        molecule["identifiers"]["molecule_hash"]
+                        for molecule in initial_molecules
+                    ]
+
+                    input_mols = client.query_molecules(molecule_hash=hashes)
+
                 except KeyError:
                     # this must be an optimisation record
                     input_mols = client.query_molecules(
-                        id=qca_record["initial_molecule"]
+                        molecule_hash=qca_record["initial_molecule"]["identifiers"][
+                            "molecule_hash"
+                        ]
                     )
-                except AttributeError:
+                except AttributeError as error:
                     raise AttributeError(
-                        "The provided client can not query molecules, make sure it is an instance of"
-                        "qcportal.client.FractalClient() with the correct address."
-                    )
+                        "The provided client failed to query molecules, make sure it is an instance of"
+                        "`qcportal.PortalClient` and pointed to the correct address."
+                    ) from error
```

Comment:
I tried a bit to cover this in a test, but had difficulties getting a bad client that would fail _in this particular way_.